### PR TITLE
[stdlib] Add `Proper` instances for some number theorems to improve rewriting with inequalities

### PR DESCRIPTION
**Pull request by haansn08**
This should add the required `Proper` instances to use `rewrite` like this:
```coq
Require Import Arith.

Definition monotonic f := forall x y, x <= y -> f x <= f y.

Goal monotonic (fun n => 5 + 6*n + 7*n^3 + Nat.sqrt(42 + 3*n)).
Proof.
  intros x y H.
  rewrite H.
  reflexivity.
Qed.
```

### CODE REVIEW

**Comment by ana-borges** [comment]
I tried out your example but it doesn't seem to be working...?

**Comment by ana-borges** [comment]
@haansn08 Cool! Could you add a changelog entry please?

Also, what about mixing `le` with `lt`, eg for `add_lt_le_mono`? Are there perhaps others missing as well?

**Review by ppedrot — CHANGES_REQUESTED** [review]
Please don't add new global instances. This goes against modularity.

**Comment by haansn08** [comment]
@ppedrot I would like to make them `#[export]`, but then `rewrite` does not find the instances anymore.

**Comment by ppedrot** [comment]
Obviously, you have to import the corresponding modules for the instances to be available. Marking them global lets you do that silently through Require, but you lose any control over the fact that the instances are there (or not).

**Comment by haansn08** [comment]
@ppedrot I have now changed the visibility to `export` as requested, but I think it will probably take a long time and nerves for a user to figure out one needs to `Import Nat` for this to work. I feel like things like that should simply work if I `Require Import Arith`.

**Comment by ppedrot** [comment]
Indeed, we need a serious set of rules for adding instances to the stdlib, so that users know when and what to import. I'm afraid that these guidelines do not exist yet...

**Comment by JasonGross** [comment]
> @ppedrot I have now changed the visibility to `export` as requested, but I think it will probably take a long time and nerves for a user to figure out one needs to `Import Nat` for this to work. I feel like things like that should simply work if I `Require Import Arith`.

The way to fix this is to add `Export Existing Instances ...` inside a module `Instances` or `Hints` or w/e in `Nat` and then `Export ` that module from `Arith`

**Comment by haansn08** [comment]
@JasonGross Thanks for the tip!

I tried putting the following at the end of `theories/Arith/PeanoNat.v`, but I get an `Error: The reference _add_lt_mono_l was not found in the current environment.` when trying to recompile.
Could somebody tell me what would be an appropriate place to put it?

```coq
(** Re-export [Proper] instances for rewriting *)

#[export] Existing Instances
  _add_lt_mono_l
  _add_lt_mono_r
  _add_lt_mono
  _add_le_mono
  _add_le_lt_mono
  _sub_le_mono_l
  _sub_le_mono_r
  _mul_lt_mono
  _mul_le_mono
  _pow_le_mono_l
  _pred_le_mono
  _le_lt_trans
  _lt_le_trans
  _log2_le_mono
  _log2_up_le_mono
  _sqrt_le_mono
  _sqrt_up_le_mono.
```

Also `make theories-arith` seems to get stuck in a loop on `ZBits.v` (it does not compile even after several hours of CPU time) when trying to compile this branch :(

**Comment by JasonGross** [comment]
> I tried putting the following at the end of `theories/Arith/PeanoNat.v`, but I get an `Error: The reference _add_lt_mono_l was not found in the current environment.` when trying to recompile.
> Could somebody tell me what would be an appropriate place to put it?

It occurs to me that there's an even better strategy, though.  You can use [filtered `Export`](https://coq.inria.fr/refman/language/core/modules.html#coq:cmd.Import) to get the hints exported.  That is, you can find all files that `Require`, e.g., `NZAddOrder` with `git grep NZAddOrder "*.v"`, and then below `Require NZAddOrder` you can add `Export (hints) NZAddOrder`.  Then you can find every file that requires one of these files, and repeat the process.  Or you can just add `Export (hints) NZAddOrder...` to all the top-level files you think people are likely to import (PeanoNat and Arith?)

**Comment by haansn08** [comment]
@ppedrot The initial example works now without any of the instances declared as `#[global]`.

@JasonGross I'm not quite sure how a filtered export could work, buit it works using `Existing Instance`, thanks!

**Comment by JasonGross** [comment]
> I'm not quite sure how a filtered export could work

I'm not sure if this is what you're asking, but typeclass instances are just hints in the `typeclass_instances` database.  What do you think wouldn't work?

**Comment by haansn08** [comment]
@JasonGross It might work, but I'm not sure how to do it. For now, `Existing Instance` seems to work fine.

**Comment by mrhaandi** [comment]
The combination of all
- `Proper (eq ==> lt ==> lt) add`
- `Proper (lt ==> eq ==> lt) add`
- `Proper (lt ==> lt ==> lt) add`
- `Proper (le ==> lt ==> lt) add`
-  ... 
is asking for poor performance in general. Especially, if it is registered on import of `PeanoNat` (which is common), it will lead to slowdowns in otherwise perfectly working developments.

**Comment by haansn08** [comment]
@mrhaandi I will gladly remove those instances, but could you explain why this would lead to poor performance?

**Comment by mrhaandi** [comment]
> @mrhaandi I will gladly remove those instances, but could you explain why this would lead to poor performance?

Having every possible combination of `le`, `lt`, `eq` as hint introduces non-determinism in proof search, which on larger examples can be detrimental. Arguably, power users of generalized rewriting would be more careful with imports. So usually it would not be problematic.
On my machine, compilation time of your branch is 9% slower than prior to the single commit. So, more performance testing is necessary here already.

**Comment by JasonGross** [comment]
For sure it is a waste to have instances like `Proper (lt ==> lt ==> lt) add` that can never be used; in order to be useful in generalized rewriting, I believe all but one of the relations on the input variables must be reflexive.  (It's possible that I'm wrong, and this will work fine when there are rewriting occurrences on both sides of the addition).  I believe there should be a default instance for addition: `Proper (eq ==> eq ==> eq) add`, and all other instances should be added manually with `Hint Extern`.  The rule of thumb I would use here is that, for the `Extern` cases, all relations which are either evars or `flip` of an evar should be instantiated with `eq`.  Then have a bunch of instances that only trigger when none of the relations are evars, one instance for every (valid) combination of `eq`, `lt`, `le`, `flip lt`, `flip le`.  (If needed, add a pass converting `flip eq`, `ge`, `gt`, `flip ge`, `flip gt` to one of these.). The code for this could look like:
```
Local Ltac instantiate_with_eq T R :=
  tryif is_evar R then unify R (@eq T) else
  lazymatch R with
  | respectful ?R1 ?R2 => instantiate_with_eq T R1; instantiate_with_eq T R2
  | flip ?R => instantiate_with_eq T R
  | _ => idtac
  end.
Local Open Scope signature_scope.
Local Ltac resolve_add R :=
  instantiate_with_eq T R;
  lazymatch R with
  | (eq ==> eq ==> eq) => exact ...
  | (eq ==> lt ==> lt) => exact ...
  | (le ==> eq ==> lt) => exact ...
  | (lt ==> lt ==> lt) => exact ...
  ...
  end.
#[export]
Hint Extern 0 (Proper ?R add) => resolve_add R : typeclass_instances.
```

**Comment by haansn08** [comment]
I now removed the `Proper` instances for the `lt` altogether. Not sure how useful they were anyways since `fun x => n * x` is only strictly monotone when `0 < n` which is not detected by rewrite. Maybe this can also be solved by using the `Hint Extern` trick.

File: theories/Arith/PeanoNat.v
Comment (Blaisorblade) [inline theories/Arith/PeanoNat.v:1183]: I suspect that could be replaced with `Export (hints) Nat.` inspired by @JasonGross 's suggestion; that would be easier to maintain.

Comment (haansn08) [reply theories/Arith/PeanoNat.v]: Works perfectly, thanks!

**Comment by JasonGross** [comment]
@coqbot bench

**Comment by JasonGross** [comment]
@coqbot run full ci

**Comment by haansn08** [comment]
@JasonGross I now squashed the commits. Is there anything else I can do to improve this PR?

**Comment by SkySkimmer** [comment]
@coq/stdlib-maintainers please make a decision instead of leaving this open for years

**Comment by andres-erbsen** [comment]
It is indeed a shame that this has been neglected so long. Who would be the most appropriate reviewer wrt considerations of using setoid rewriting at scale? I personally feel compelled to defer to @JasonGross, but I know he doesn't have all that much extra time on his hands these days and perhaps there's somebody who's actually in charge of setoid rewrite and associated infrastructure? @coq/typeclasses-maintainers ?

**Review by JasonGross — APPROVED** [review]
I think the optimal proper instances have eq in all argument slots but one, but these are good and there's no reason to block this PR.  Can someone rebase and let's check full CI and bench again, and if everything looks fine let's merge

**Comment by andres-erbsen** [comment]
The PR rebeases cleanly but I get an error on push. Might it be that I am missing yet another type of permission, or would the PR author have to do it?

**Comment by JasonGross** [comment]
I tensed through the web interface by temporarily enabling it

@coqbot bench

**Comment by andres-erbsen** [comment]
@coqbot run full ci

**Comment by andres-erbsen** [comment]
hb failure is unrelated (`Error: Unbound type constructor DeclareInd.default_dep_elim`)

**Comment by andres-erbsen** [comment]
I intend to merge on Monday or when I next look at this PR.

**Comment by proux01** [comment]
Ping @andres-erbsen

**Pull request closed without merging**
